### ISSUE TITLE
redundant_clone: false positive in equality comparison

### ISSUE THREAD

**Issue by ebroto** [rust-lang/rust-clippy#5700]
```rust
#[derive(PartialEq)]
struct Foo;

struct Bar;

impl ToOwned for Bar {
    type Owned = Foo;
    fn to_owned(&self) -> Foo {
        Foo {}
    }
}

impl std::borrow::Borrow<Bar> for Foo {
    fn borrow(&self) -> &Bar {
        static BAR: Bar = Bar {};
        &BAR
    }
}

fn main() {
    let f = Foo {};
    let b = Bar {};

    if f == b.to_owned() {
        println!("they match")
    }
}
```
([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=954f7d65255b9a52fc1dfa41c42b0dfb))

Note that there is no `PartialEq` implementation between Foo and Bar. This generates the following warning:

```
warning: redundant clone
  --> src/main.rs:24:14
   |
24 |     if f == b.to_owned() {
   |              ^^^^^^^^^^^ help: remove this
   |
   = note: `#[warn(clippy::redundant_clone)]` on by default
note: this value is dropped without further use
  --> src/main.rs:24:13
   |
24 |     if f == b.to_owned() {
   |             ^
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#redundant_clone

warning: 1 warning emitted
```

But applying the suggestion causes this error

```
error[E0308]: mismatched types
  --> src/main.rs:24:13
   |
24 |     if f == b {
   |             ^ expected struct `Foo`, found struct `Bar`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0308`.
```
([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=dfe489af91e9d9044b24f378a8480f22))

Clippy version: 0.0.212

**Comment by flip1995** [comment]
I don't see this as a bug in `redundant_clone`. The `to_owned` method is redundant, since `Bar` does not need to be cloned to be used. 

Instead the `to_owned` method is used to convert a type in this example. For this, `From<Bar> for Foo` should be implemented.

I think this is a constructed example, that is not realistic in real code.

**Comment by ebroto** [comment]
Yep I agree the code is not ideal, I ran into this when testing edge cases of `cmp_owned` when working in #4874.

I opened the issue because the suggestion causes a compilation error and because I understood that the temporary is actually being used in the comparison.

But maybe it's not worth the effort because it's unlikely that someone writes this? In fact having this linted by `redundant_clone` could be a hint that this should be changed, even though the suggestion is not that explicit.

**Comment by ebroto** [comment]
To be clear, I assumed that if it results in a compilation error then it's a false positive, but maybe this is unlikely to happen

**Comment by flip1995** [comment]
> I assumed that if it results in a compilation error then it's a false positive

Normally it would be, but in this case `to_owned` is misused for type conversion. So I'm not sure if anything has to be changed in the `redundant_clone` lint. But I would leave this open, if someone else somehow experiences this.

**Comment by profetia** [comment]
I can no longer reproduce this in the latest nightly.

**Comment by samueltardieu** [comment]
> I can no longer reproduce this in the latest nightly.

I can. Are you sure you have activated the `clippy::redundant_clone` lint which is allow by default?

**Comment by profetia** [comment]
Wiered. Im using `clippy::all` in the [playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=82b79d28ee29c5592df7ebeb44deaec2)

**Comment by y21** [comment]
> Wiered. Im using `clippy::all` in the [playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=82b79d28ee29c5592df7ebeb44deaec2)

`clippy::all` only includes the lints that are enabled by default, so `#[warn(clippy::all)]` is really a no-op (well, it also "downgrades" deny by default lints to warn by default, but doesn't enable anything else). Just today [a thread was created on zulip](https://rust-lang.zulipchat.com/#narrow/channel/257328-clippy/topic/Renaming.20clippy.3A.3Aall/with/510839233) with the idea to rename this because it's confusing.

Anyway, with `#![warn(clippy::nursery)]` (or just `#![warn(clippy::redundant_clone)]`) it indeed still repros.
